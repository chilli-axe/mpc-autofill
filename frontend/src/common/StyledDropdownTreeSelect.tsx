import DropdownTreeSelect from "react-dropdown-tree-select";
import styled from "styled-components";

export const StyledDropdownTreeSelect = styled(DropdownTreeSelect)`
  .tag {
    color: black;
    background-color: #dddddd;
  }
  .tag-remove {
    color: #666666;
  }

  .dropdown-trigger {
    border-radius: 0.25rem;
    background-color: white;
  }
  .dropdown-content {
    border-radius: 0.25rem;
  }

  .search {
    background-color: white;
    color: black;
  }
  .search::placeholder {
    color: black;
  }
  .placeholder {
    color: gray;
  }

  .toggle {
    font: normal normal normal 12px/1 bootstrap-icons;
    top: 2px;
    left: 2px;
  }

  .toggle.collapsed::after {
    content: "\F4FA";
  }

  .toggle.expanded::after {
    content: "\F2E6";
  }

  color: black;

  .root {
    padding: 0;
    margin: 0;
  }
`;
