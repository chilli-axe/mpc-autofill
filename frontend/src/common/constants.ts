import { CardType as CardTypeSchema } from "@/common/schema_types";
import { Cardstock, CardType, Faces, SortBy } from "@/common/types";
export const ProjectName = "MPC Autofill";
export const MakePlayingCards = "MakePlayingCards.com";
export const MakePlayingCardsURL = "https://www.makeplayingcards.com";

export const Card: CardType = CardTypeSchema.Card;
export const Cardback: CardType = CardTypeSchema.Cardback;
export const Token: CardType = CardTypeSchema.Token;

export const SelectedImageSeparator = "@";
export const CardTypeSeparator = ":";
export const FaceSeparator = "//";

export const CardTypePrefixes: { [prefix: string]: CardType } = {
  "": Card,
  b: Cardback,
  t: Token,
};

export const ReversedCardTypePrefixes = Object.fromEntries(
  Object.keys(CardTypePrefixes).map((prefix: string) => [
    CardTypePrefixes[prefix],
    prefix.length > 0 ? prefix + CardTypeSeparator : prefix,
  ])
);

export const Front: Faces = "front";
export const Back: Faces = "back";

export const ToggleButtonHeight = 38; // pixels
export const NavbarHeight = 50; // pixels - aligns with the natural height of the navbar
export const RibbonHeight = 54; // pixels
export const NavbarLogoHeight = 40; // pixels
export const ContentMaxWidth = 1200; // pixels - aligns with bootstrap's large breakpoint

export const MinimumDPI = 0;
export const MaximumDPI = 1500;
export const DPIStep = 50;

export const MaximumSize = 30; // megabytes
export const SizeStep = 1;

export const CSRFKey = "csrftoken";
export const SearchSettingsKey = "searchSettings";
export const GoogleAnalyticsConsentKey = "googleAnalyticsConsent";
export const BackendURLKey = "backendURL";

export const Brackets: Array<number> = [
  18, 36, 55, 72, 90, 108, 126, 144, 162, 180, 198, 216, 234, 396, 504, 612,
];

export const ProjectMaxSize: number = Brackets[Brackets.length - 1];

export enum QueryTags {
  BackendSpecific = "backendSpecific",
  SearchResults = "searchResults",
  SampleCards = "sampleCards",
}

export const S27: Cardstock = "(S27) Smooth";
export const S30: Cardstock = "(S30) Standard Smooth";
export const S33: Cardstock = "(S33) Superior Smooth";
export const M31: Cardstock = "(M31) Linen";
export const P10: Cardstock = "(P10) Plastic";
export const Cardstocks: Array<Cardstock> = [S27, S30, S33, M31, P10];

export const CardstockFoilCompatibility: { [cardstock in Cardstock]: boolean } =
  {
    [S27]: true,
    [S30]: true,
    [S33]: true,
    [M31]: true,
    [P10]: false,
  };

export const GoogleDriveImageAPIURL =
  "https://script.google.com/macros/s/AKfycbw8laScKBfxda2Wb0g63gkYDBdy8NWNxINoC4xDOwnCQ3JMFdruam1MdmNmN4wI5k4/exec";

export const SearchResultsEndpointPageSize = 300;
export const CardEndpointPageSize = 1000;

export enum CSVHeaders {
  quantity = "Quantity",
  frontQuery = "Front",
  frontSelectedImage = "Front ID",
  backQuery = "Back",
  backSelectedImage = "Back ID",
}

export const ExploreDebounceMS = 700;
export const ExplorePageSize = 60;

export const SortByOptions: { [option in SortBy]: string } = {
  dateCreatedDescending: "Date Created (Newest-Oldest)",
  dateCreatedAscending: "Date Created (Oldest-Newest)",
  dateModifiedDescending: "Date Modified (Newest-Oldest)",
  dateModifiedAscending: "Date Modified (Oldest-Newest)",
  nameAscending: "Name (A-Z)",
  nameDescending: "Name (Z-A)",
};
