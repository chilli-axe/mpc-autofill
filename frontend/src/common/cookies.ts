/**
 * Data API for interacting with anything stored in cookies or local storage.
 */

// import Ajv2020 from "ajv/dist/2020";
import Cookies from "js-cookie";

import {
  BackendURLKey,
  CSRFKey,
  GoogleAnalyticsConsentKey,
  SearchSettingsKey,
} from "@/common/constants";
import { Convert } from "@/common/schema_types";
import { SearchSettings, SourceDocuments, SourceRow } from "@/common/types";
import { getSourceRowsFromSourceSettings } from "@/common/utils";
import { getDefaultSearchSettings } from "@/store/slices/searchSettingsSlice";

// import * as SearchSettingsSchema from "../../../schemas/schemas/SearchSettings.json";

// const ajv = new Ajv2020();

//# region CSRF
// TODO: unsure if we still need this.

export function getCSRFHeader(): HeadersInit | undefined {
  const csrfToken = Cookies.get(CSRFKey);
  if (csrfToken != null) {
    return { "X-CSRFToken": csrfToken };
  }
  return undefined;
}

//# endregion

//# region search settings

/**
 * Get search settings from localStorage data. If valid data is retrieved,
 * ensure that all `sourceDocuments` are included in the returned settings,
 * with any new sources that weren't previously included added to the end and enabled.
 */
export function getLocalStorageSearchSettings(
  sourceDocuments: SourceDocuments
): SearchSettings {
  const serialisedRawSettings = localStorage.getItem(SearchSettingsKey) ?? "{}";
  try {
    const searchSettings = Convert.toSearchSettings(serialisedRawSettings);
    // great, the user has valid search settings stored in their browser local storage.
    // reconcile against sourceDocuments
    const sourceInDatabaseSet: Set<number> = new Set(
      Object.values(sourceDocuments).map((sourceDocument) =>
        parseInt(sourceDocument.pk)
      )
    );
    // types have to be narrowed here because quicktype doesn't support our SourceRow data structure :(
    const sources: Array<SourceRow> = getSourceRowsFromSourceSettings(
      searchSettings.sourceSettings
    );
    const sourceInLocalStorageSet: Set<number> = new Set(
      sources.map((row) => row[0])
    );
    // one fat line of reconciliation, good luck reading this future nick! i wrote this at 12:26am.
    searchSettings.sourceSettings.sources = sources
      .filter((row: SourceRow) => sourceInDatabaseSet.has(row[0]))
      .concat(
        Array.from(sourceInDatabaseSet)
          .filter((pk: number) => !sourceInLocalStorageSet.has(pk))
          .map((pk: number) => [pk, true])
      );
    return searchSettings;
  } catch (e) {
    // quicktype will throw an error if the user's stored settings do not match the schema
    // e.g. upon first page load
    // just return the default settings
    return getDefaultSearchSettings(sourceDocuments);
  }
}

export function setLocalStorageSearchSettings(settings: SearchSettings): void {
  localStorage.setItem(SearchSettingsKey, JSON.stringify(settings));
}

//# endregion

//# region google analytics consent

export function getGoogleAnalyticsConsent(): boolean | undefined {
  const rawConsent = Cookies.get(GoogleAnalyticsConsentKey);
  return rawConsent != undefined ? JSON.parse(rawConsent) === true : undefined;
}

export function setGoogleAnalyticsConsent(consent: boolean): void {
  Cookies.set(GoogleAnalyticsConsentKey, JSON.stringify(consent), {
    expires: 365,
    sameSite: "strict",
  });
}

//# endregion

//# region backend

export function getLocalStorageBackendURL() {
  return localStorage.getItem(BackendURLKey);
}

export function setLocalStorageBackendURL(url: string) {
  localStorage.setItem(BackendURLKey, url);
}

export function clearLocalStorageBackendURL(): void {
  localStorage.removeItem(BackendURLKey);
}

//# endregion
