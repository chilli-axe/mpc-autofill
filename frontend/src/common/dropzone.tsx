/**
 * A small wrapper around the `react-dropzone` Dropzone for consistent presentation.
 */

import React, { useCallback } from "react";
import { DropzoneRootProps, useDropzone } from "react-dropzone";
import styled from "styled-components";

import { useAppDispatch } from "@/common/types";
import { setNotification } from "@/store/slices/toastsSlice";

const getColor = (props: DropzoneRootProps) => {
  if (props.isDragAccept) {
    return "#00e676";
  }
  if (props.isDragReject) {
    return "#ff1744";
  }
  if (props.isFocused) {
    return "#2196f3";
  }
  return "#eeeeee";
};

const Container = styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
  border-width: 2px;
  border-radius: 2px;
  border-color: ${(props: DropzoneRootProps) => getColor(props)};
  border-style: dashed;
  color: #bdbdbd;
  outline: none;
  transition: border 0.24s ease-in-out;
  cursor: pointer;
`;

interface StyledDropzoneProps {
  mimeTypes: { [mimeType: string]: Array<string> };
  fileUploadCallback: {
    (fileContent: string | ArrayBuffer | null): void;
  };
  label: string;
  disabled: boolean;
}

export function TextFileDropzone({
  mimeTypes,
  fileUploadCallback,
  label,
  disabled,
}: StyledDropzoneProps) {
  const dispatch = useAppDispatch();

  // let the user know when something goes wrong with reading their file
  const onAbort = useCallback(
    () =>
      dispatch(
        setNotification([
          "dropzone-onabort",
          {
            name: "File upload error",
            message: "File reading was stopped.",
            level: "error",
          },
        ])
      ),
    [dispatch]
  );
  const onError = useCallback(
    () =>
      dispatch(
        setNotification([
          "dropzone-onerror",
          {
            name: "File upload error",
            message: "An error occurred while reading the file.",
            level: "error",
          },
        ])
      ),
    [dispatch]
  );
  const onDrop = useCallback(
    (acceptedFiles: Array<File>) => {
      acceptedFiles.forEach((file: File) => {
        const reader = new FileReader();
        reader.onabort = onAbort;
        reader.onerror = onError;
        reader.onload = () => {
          fileUploadCallback(reader.result);
        };
        reader.readAsText(file);
      });
    },
    [fileUploadCallback, onAbort, onError]
  );

  const { getRootProps, getInputProps, isFocused, isDragAccept, isDragReject } =
    useDropzone({ onDrop, accept: mimeTypes, maxFiles: 1, disabled });

  return (
    <div className="container">
      <Container
        {...getRootProps({ isFocused, isDragAccept, isDragReject })}
        aria-label={label ?? "dropzone"}
      >
        <input {...getInputProps()} />
        Drag and drop a file here, or click to select a file.
      </Container>
    </div>
  );
}
