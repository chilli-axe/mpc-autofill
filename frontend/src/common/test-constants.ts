/**
 * Some pre-built objects which can be used to build up Redux state for tests.
 */

import { Card, MaximumDPI, MaximumSize, MinimumDPI } from "@/common/constants";
import { CardType as CardTypeSchema, SourceType } from "@/common/schema_types";
import {
  BackendState,
  CardDocument,
  Project,
  SearchSettings,
  SourceDocument,
  SourceDocuments,
} from "@/common/types";

//# region backend

export const localBackendURL = "https://127.0.0.1:8000";
export const localBackend: BackendState = { url: localBackendURL };
export const noBackend: BackendState = { url: null };

//# endregion

//# region sources

export const sourceDocument1: SourceDocument = {
  pk: 0,
  key: "source_1",
  name: "Source 1",
  sourceType: SourceType.GoogleDrive,
  externalLink: undefined,
  description: "",
};

export const sourceDocument2: SourceDocument = {
  pk: 1,
  key: "source_2",
  name: "Source 2",
  sourceType: SourceType.GoogleDrive,
  externalLink: undefined,
  description: "",
};

export const sourceDocument3: SourceDocument = {
  pk: 2,
  key: "source_3",
  name: "Source 3",
  sourceType: SourceType.GoogleDrive,
  externalLink: undefined,
  description: "",
};

export const sourceDocument4: SourceDocument = {
  pk: 3,
  key: "source_4",
  name: "Source 4",
  sourceType: SourceType.GoogleDrive,
  externalLink: undefined,
  description: "",
};

export const sourceDocuments: SourceDocuments = {
  [sourceDocument1.pk]: sourceDocument1,
  [sourceDocument2.pk]: sourceDocument2,
  [sourceDocument3.pk]: sourceDocument3,
  [sourceDocument4.pk]: sourceDocument4,
};

//# endregion

//# region cards

export const cardDocument1: CardDocument = {
  identifier: "abc123",
  cardType: CardTypeSchema.Card,
  name: "Card 1",
  priority: 0,
  source: sourceDocument1.key,
  sourceName: sourceDocument1.name,
  sourceId: sourceDocument1.pk,
  sourceVerbose: sourceDocument1.name,
  sourceType: SourceType.GoogleDrive,
  sourceExternalLink: undefined,
  dpi: 1200,
  searchq: "card 1",
  extension: "png",
  dateCreated: "1st January, 2000", // formatted by backend
  dateModified: "1st January, 2000", // formatted by backend
  downloadLink: "",
  size: 10_000_000,
  smallThumbnailUrl: "",
  mediumThumbnailUrl: "",
  language: "EN",
  tags: [],
};

export const cardDocument2: CardDocument = {
  identifier: "abc1234",
  cardType: CardTypeSchema.Card,
  name: "Card 2",
  priority: 0,
  source: sourceDocument1.key,
  sourceName: sourceDocument1.name,
  sourceId: sourceDocument1.pk,
  sourceVerbose: sourceDocument1.name,
  sourceType: SourceType.GoogleDrive,
  sourceExternalLink: undefined,
  dpi: 1200,
  searchq: "card 2",
  extension: "png",
  dateCreated: "1st January, 2000", // formatted by backend
  dateModified: "1st January, 2000", // formatted by backend
  downloadLink: "",
  size: 10_000_000,
  smallThumbnailUrl: "",
  mediumThumbnailUrl: "",
  language: "EN",
  tags: [],
};

export const cardDocument3: CardDocument = {
  identifier: "abc12345",
  cardType: CardTypeSchema.Card,
  name: "Card 3",
  priority: 0,
  source: sourceDocument1.key,
  sourceName: sourceDocument1.name,
  sourceId: sourceDocument1.pk,
  sourceVerbose: sourceDocument1.name,
  sourceType: SourceType.GoogleDrive,
  sourceExternalLink: undefined,
  dpi: 1200,
  searchq: "card 3",
  extension: "png",
  dateCreated: "1st January, 2000", // formatted by backend
  dateModified: "1st January, 2000", // formatted by backend
  downloadLink: "",
  size: 10_000_000,
  smallThumbnailUrl: "",
  mediumThumbnailUrl: "",
  language: "EN",
  tags: [],
};

export const cardDocument4: CardDocument = {
  identifier: "abc123456",
  cardType: CardTypeSchema.Card,
  name: "Card 4",
  priority: 0,
  source: sourceDocument1.key,
  sourceName: sourceDocument1.name,
  sourceId: sourceDocument1.pk,
  sourceVerbose: sourceDocument1.name,
  sourceType: SourceType.GoogleDrive,
  sourceExternalLink: undefined,
  dpi: 1200,
  searchq: "card 4",
  extension: "png",
  dateCreated: "1st January, 2000", // formatted by backend
  dateModified: "1st January, 2000", // formatted by backend
  downloadLink: "",
  size: 10_000_000,
  smallThumbnailUrl: "",
  mediumThumbnailUrl: "",
  language: "EN",
  tags: [],
};

export const cardDocument5: CardDocument = {
  identifier: "abc1234567",
  cardType: CardTypeSchema.Cardback,
  name: "Card 5",
  priority: 0,
  source: sourceDocument2.key,
  sourceName: sourceDocument2.name,
  sourceId: sourceDocument2.pk,
  sourceVerbose: `${sourceDocument2.name} Cardbacks`,
  sourceType: SourceType.GoogleDrive,
  sourceExternalLink: undefined,
  dpi: 1200,
  searchq: "card 5",
  extension: "png",
  dateCreated: "1st January, 2000", // formatted by backend
  dateModified: "1st January, 2000", // formatted by backend
  downloadLink: "",
  size: 10_000_000,
  smallThumbnailUrl: "",
  mediumThumbnailUrl: "",
  language: "EN",
  tags: [],
};

export const cardDocument6: CardDocument = {
  identifier: "abc12345678",
  cardType: CardTypeSchema.Token,
  name: "Card 6",
  priority: 0,
  source: sourceDocument3.key,
  sourceName: sourceDocument3.name,
  sourceId: sourceDocument3.pk,
  sourceVerbose: `${sourceDocument3.name} Tokens`,
  sourceType: SourceType.GoogleDrive,
  sourceExternalLink: undefined,
  dpi: 1200,
  searchq: "card 6",
  extension: "png",
  dateCreated: "1st January, 2000", // formatted by backend
  dateModified: "1st January, 2000", // formatted by backend
  downloadLink: "",
  size: 10_000_000,
  smallThumbnailUrl: "",
  mediumThumbnailUrl: "",
  language: "EN",
  tags: [],
};

//# endregion

//# region project

export const projectSelectedImage1: Project = {
  members: [
    {
      front: {
        query: { query: "my search query", cardType: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
  ],
  cardback: null,
  mostRecentlySelectedSlot: null,
};

export const projectThreeMembersSelectedImage1: Project = {
  members: [
    {
      front: {
        query: { query: "my search query", cardType: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
    {
      front: {
        query: { query: "my search query", cardType: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
    {
      front: {
        query: { query: "my search query", cardType: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
  ],
  cardback: null,
  mostRecentlySelectedSlot: null,
};

export const projectSelectedImage2: Project = {
  members: [
    {
      front: {
        query: { query: "my search query", cardType: Card },
        selectedImage: cardDocument2.identifier,
        selected: false,
      },
      back: null,
    },
  ],
  cardback: null,
  mostRecentlySelectedSlot: null,
};

//# endregion

//# region search settings

export const defaultSettings: SearchSettings = {
  searchTypeSettings: { fuzzySearch: false, filterCardbacks: false },
  sourceSettings: {
    sources: [
      [sourceDocument1.pk, true],
      [sourceDocument2.pk, true],
      [sourceDocument3.pk, true],
      [sourceDocument4.pk, true],
    ],
  },
  filterSettings: {
    minimumDPI: MinimumDPI,
    maximumDPI: MaximumDPI,
    maximumSize: MaximumSize,
    languages: [],
    includesTags: [],
    excludesTags: ["NSFW"],
  },
};

//# endregion
