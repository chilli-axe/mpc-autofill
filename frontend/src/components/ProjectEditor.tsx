/**
 * This component represents the complete MPC Autofill project editor, ready to
 * drop into a page (as the only component). Must be wrapped with a Redux provider.
 */

import React, { useEffect } from "react";
import Col from "react-bootstrap/Col";
import Row from "react-bootstrap/Row";
import styled from "styled-components";

import { RibbonHeight } from "@/common/constants";
import { useAppSelector } from "@/common/types";
import { NoBackendDefault } from "@/components/NoBackendDefault";
import { OverflowCol } from "@/components/OverflowCol";
import { Ribbon } from "@/components/Ribbon";
import { SelectedImagesRibbon } from "@/features/bulkManagement/SelectedImagesRibbon";
import { CardGrid } from "@/features/card/CardGrid";
import { CommonCardback } from "@/features/card/CommonCardback";
import { Export } from "@/features/export/Export";
import { FinishedMyProject } from "@/features/export/FinishedMyProjectModal";
import { FinishSettings } from "@/features/finishSettings/FinishSettings";
import { Import } from "@/features/import/Import";
import { SearchSettings } from "@/features/searchSettings/SearchSettings";
import { Status } from "@/features/status/Status";
import { useBackendConfigured } from "@/store/slices/backendSlice";
import {
  selectIsProjectEmpty,
  selectProjectCardback,
} from "@/store/slices/projectSlice";

function ProjectEditor() {
  // TODO: should we periodically ping the backend to make sure it's still alive?
  //# region queries and hooks

  const backendConfigured = useBackendConfigured();
  const cardback = useAppSelector(selectProjectCardback);
  const isProjectEmpty = useAppSelector(selectIsProjectEmpty);

  //# endregion

  //# region effects

  /**
   * Ask the user for confirmation before they close the page if their project has any cards in it.
   */
  useEffect(() => {
    const handler = (event: BeforeUnloadEvent) => {
      if (!isProjectEmpty) {
        event.preventDefault();
        return false;
      }
    };
    window.addEventListener("beforeunload", handler);
    return () => {
      window.removeEventListener("beforeunload", handler);
    };
  }, [isProjectEmpty]);

  //# endregion

  return (
    <>
      {backendConfigured ? (
        <Row className="g-0">
          <Ribbon className="g-0">
            <SelectedImagesRibbon />
          </Ribbon>
          <OverflowCol
            lg={8}
            md={8}
            sm={6}
            xs={6}
            data-testid="left-panel"
            heightDelta={RibbonHeight}
          >
            <CardGrid />
          </OverflowCol>
          <OverflowCol
            data-testid="right-panel"
            lg={4}
            md={4}
            sm={6}
            xs={6}
            style={{ zIndex: 1 }}
            className="px-2"
            heightDelta={RibbonHeight}
          >
            <Status />
            <Row className="g-0 pt-2">
              <FinishSettings />
            </Row>
            <Row className="g-0 pt-2">
              <SearchSettings />
            </Row>
            <Row className="g-0 pt-2">
              <Col lg={7} md={12} sm={12} xs={12}>
                <Import />
              </Col>
              <Col lg={5} md={12} sm={12} xs={12}>
                <Export />
              </Col>
              <FinishedMyProject />
            </Row>
            <Col className="g-0 pt-2" lg={{ span: 8, offset: 2 }} md={12}>
              <CommonCardback selectedImage={cardback} />
            </Col>
          </OverflowCol>
        </Row>
      ) : (
        <NoBackendDefault />
      )}
    </>
  );
}

export default ProjectEditor;
