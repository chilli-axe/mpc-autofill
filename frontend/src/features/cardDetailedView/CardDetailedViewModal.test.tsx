import { screen, waitFor } from "@testing-library/react";

import { Card, Front } from "@/common/constants";
import { cardDocument1 } from "@/common/test-constants";
import {
  expectCardGridSlotState,
  renderWithProviders,
} from "@/common/test-utils";
import ProjectEditor from "@/components/ProjectEditor";
import {
  cardDocumentsOneResult,
  defaultHandlers,
  searchResultsOneResult,
  sourceDocumentsOneResult,
} from "@/mocks/handlers";
import { server } from "@/mocks/server";

//# region snapshot tests

test("the html structure of a CardDetailedViewModal", async () => {
  server.use(
    cardDocumentsOneResult,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, {
    preloadedState: {
      project: {
        members: [
          {
            front: {
              query: { query: "my search query", cardType: Card },
              selectedImage: cardDocument1.identifier,
              selected: false,
            },
            back: null,
          },
        ],
        cardback: null,
        mostRecentlySelectedSlot: null,
      },
    },
  });
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);

  screen.getByAltText(cardDocument1.name).click();
  await waitFor(() => expect(screen.getByText("Card Details")));
  await waitFor(() => expect(screen.getByText("English")));

  expect(screen.getByTestId("detailed-view")).toMatchSnapshot();
});

//# endregion
