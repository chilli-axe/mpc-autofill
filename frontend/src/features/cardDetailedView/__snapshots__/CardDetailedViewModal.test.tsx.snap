// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the html structure of a CardDetailedViewModal 1`] = `
.c2 {
  max-width: 100%;
  overflow-x: scroll;
}

.c3 {
  table-layout: auto;
}

.c4 {
  width: 0%;
  text-align: left;
}

.c5 {
  width: 0%;
  text-align: left;
}

.c6 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: solid 1px #ffffff00;
  -webkit-transition: outline 0.2s ease-in-out;
  transition: outline 0.2s ease-in-out;
}

.c6:hover {
  outline-color: #ffffffff;
  cursor: pointer;
}

.c7 {
  padding-right: 0.5em;
}

.c1 {
  z-index: 1;
  opacity: 1;
}

.c1:hover {
  cursor: auto;
}

.c0 {
  z-index: 0;
  background: #4e5d6c;
  border: solid 0px black;
}

<div
  class="modal-dialog modal-xl"
  data-testid="detailed-view"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        Card Details
      </div>
      <button
        aria-label="Close"
        class="btn-close"
        type="button"
      />
    </div>
    <div
      class="modal-body"
    >
      <div
        class="row"
      >
        <div
          class="col-lg-5 mb-3 mb-lg-0"
          style="position: relative;"
        >
          <div
            class="c0 rounded-xl shadow-lg ratio ratio-7x5"
          >
            <img
              alt="Card 1"
              class="c1 card-img card-img-fade-in"
              data-nimg="fill"
              decoding="async"
              loading="lazy"
              src=""
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
            />
          </div>
        </div>
        <div
          class="col-lg-7"
        >
          <h4>
            Card 1
          </h4>
          <div
            class="c2"
          >
            <table
              class="c3 table table-hover"
            >
              <tbody>
                <tr>
                  <th
                    class="c4"
                    scope="row"
                  >
                    Source Name
                  </th>
                  <td
                    class="c5"
                  >
                    Source 1
                  </td>
                </tr>
                <tr>
                  <th
                    class="c4"
                    scope="row"
                  >
                    Source Type
                  </th>
                  <td
                    class="c5"
                  >
                    Google Drive
                  </td>
                </tr>
                <tr>
                  <th
                    class="c4"
                    scope="row"
                  >
                    Class
                  </th>
                  <td
                    class="c5"
                  >
                    Card
                  </td>
                </tr>
                <tr>
                  <th
                    class="c4"
                    scope="row"
                  >
                    Identifier
                  </th>
                  <td
                    class="c5"
                  >
                    <code
                      class="c6"
                    >
                      abc123
                    </code>
                  </td>
                </tr>
                <tr>
                  <th
                    class="c4"
                    scope="row"
                  >
                    Language
                  </th>
                  <td
                    class="c5"
                  >
                    English
                  </td>
                </tr>
                <tr>
                  <th
                    class="c4"
                    scope="row"
                  >
                    Tags
                  </th>
                  <td
                    class="c5"
                  >
                    Untagged
                  </td>
                </tr>
                <tr>
                  <th
                    class="c4"
                    scope="row"
                  >
                    Resolution
                  </th>
                  <td
                    class="c5"
                  >
                    1200 DPI
                  </td>
                </tr>
                <tr>
                  <th
                    class="c4"
                    scope="row"
                  >
                    Date Created
                  </th>
                  <td
                    class="c5"
                  >
                    1st January, 2000
                  </td>
                </tr>
                <tr>
                  <th
                    class="c4"
                    scope="row"
                  >
                    Date Modified
                  </th>
                  <td
                    class="c5"
                  >
                    1st January, 2000
                  </td>
                </tr>
                <tr>
                  <th
                    class="c4"
                    scope="row"
                  >
                    File Size
                  </th>
                  <td
                    class="c5"
                  >
                    10 MB
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
          <div
            class="d-grid gap-0"
          >
            <button
              class="btn btn-primary"
              type="button"
            >
              <i
                class="c7 bi bi-cloud-arrow-down"
              />
               
              Download Image
            </button>
          </div>
          <form
            class=""
          >
            <div
              class="g-0 pt-3 row"
            >
              <div
                class="col-lg-2 col-md-3"
              >
                <input
                  alt="Quantity of card to add to project"
                  class="form-control"
                  max="611"
                  min="1"
                  pattern="[0-9]*"
                  placeholder="1"
                  type="number"
                  value="1"
                />
              </div>
              <div
                class="col-lg-10 col-md-9"
              >
                <div
                  class="d-grid gap-0"
                >
                  <button
                    class="btn btn-success"
                    type="submit"
                  >
                    <i
                      class="c7 bi bi-plus-circle"
                    />
                     Add to Project
                  </button>
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-secondary"
        type="button"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;
